codec-http2/src/main/java/io/netty/handler/codec/http2/AbstractHttp2StreamChannel.java

```diff
@@ -324,11 +351,143 @@ private boolean doRead0(Object msg, RecvByteBufAllocator.Handle allocHandle) {
         return true;
     }
 
-    private final class Unsafe extends AbstractUnsafe {
+    private void reevaluateWritability() {
+        ChannelOutboundBuffer buffer = unsafe().outboundBuffer();
+        // If the buffer is not writable but should be writable, then write and flush a dummy object
+        // to trigger a writability change.
+        if (!buffer.isWritable() && buffer.totalPendingWriteBytes() < config.getWriteBufferHighWaterMark()) {
+            unsafe().outboundBuffer().addMessage(REEVALUATE_WRITABILITY_MESSAGE, 1, voidPromise());
+            unsafe().flush();
+        }
+    }
+
+    private static int dataFrameFlowControlBytes(Http2DataFrame frame) {
+        return frame.content().readableBytes()
+               + frame.padding()
+               // +1 to account for the pad length field. See http://httpwg.org/specs/rfc7540.html#DATA
```

Comment:
> The +1 is already accounted for in padding(); a value of 1 should result with PADDED == true and Pad Length == 0. Also, &1 would just make sure that the result is even...

I think it might not have been accounted for back when this code was written. I remember the HTTP/2 codec having had this bug, where the presence of the padding length field was not accounted for by flow control.

> TODO: But this probably does show we need a convenience method like Http2DataFrame.flowControlledBytes()

+1